compiler/passes/normalize.cpp

```diff
@@ -1293,14 +1301,163 @@ static bool isVoidReturn(CallExpr* call) {
   return retval;
 }
 
-static void insertRetMove(FnSymbol* fn, VarSymbol* retval, CallExpr* ret) {
+static bool hasGenericArrayReturn(FnSymbol* fn) {
+  if (returnsArray(fn)) {
+    BlockStmt* typeExpr = fn->retExprType;
+
+    CallExpr* call = toCallExpr(typeExpr->body.tail);
+    int nArgs = call->numActuals();
+    Expr* domExpr = call->get(1);
+    Expr* eltExpr = nArgs == 2 ? call->get(2) : NULL;
+    bool noDom = (isSymExpr(domExpr) && toSymExpr(domExpr)->symbol() == gNil);
+
+    if (noDom || eltExpr == NULL) {
+      // Either the domain is not provided explicitly as part of the return
+      // type, or the element type is not provided, or both
+
+      return true;
+    }
+  }
+
+  return false;
+}
+
+static void insertDomainCheck(Expr* actualRet, CallExpr* retVar,
+                              Expr* domExpr) {
+  VarSymbol* unrefTemp = newTemp();
+  retVar->insertBefore(new DefExpr(unrefTemp));
+  CallExpr* makeArray = new CallExpr(PRIM_MOVE, unrefTemp,
+                                     new CallExpr("chpl__unref",
+                                                  actualRet->copy()));
+  retVar->insertBefore(makeArray);
+
+  VarSymbol* actualDom = newTemp();
+  retVar->insertBefore(new DefExpr(actualDom));
+  CallExpr* exprDom = new CallExpr("_dom", gMethodToken, unrefTemp);
+  retVar->insertBefore(new CallExpr(PRIM_MOVE, actualDom, exprDom));
+
+  CallExpr* checkDom = new CallExpr("==", actualDom, domExpr->copy());
+
+  VarSymbol* cond = newTemp();
+  retVar->insertBefore(new DefExpr(cond));
+
+  VarSymbol* checkVar = newTemp();
+  retVar->insertBefore(new DefExpr(checkVar));
+  retVar->insertBefore(new CallExpr(PRIM_MOVE, checkVar, checkDom));
+  retVar->insertBefore(new CallExpr(PRIM_MOVE, cond,
+                                    new CallExpr("_cond_test", checkVar)));
+
+  Symbol* msg = new_StringSymbol("domain mismatch on return");
+  BlockStmt* failureMode = new BlockStmt(new CallExpr("halt",
```

Comment:
I think it would simplify this code if you just call a new module routine like `chpl_check_domains_match` or something. That routine should check if bounds checking is on (ignoring the error if it's not on) and should ideally halt using one of the methods Elliot added in #10191 - boundsCheckHalt().